lib/Parse/ParseStmt.cpp

```diff
@@ -1793,7 +1793,8 @@ Parser::parseStmtConditionElement(SmallVectorImpl<StmtConditionElement> &result,
     auto declRefExpr = new (Context) UnresolvedDeclRefExpr(bindingName,
                                                            DeclRefKind::Ordinary,
                                                            loc);
-    
+    // TODO(concurrency): Consider marking this with a flag so
+    //  diagnoseUnhandledAsyncSite has an easier time detecting this situation?
```

Comment:
I'd like some opinion on this - should we add a new bit for "this is an expression made up for the shorthand if let" or something more generic we could call this?

OR, what if we removed the `setImplicit()` or do we want to definitely keep that flag since it is true that we "made up" this expression?

Reply:
You said that 
> and that the 'diagnoseUnhandledAsyncSite' explicitly avoids reporting errors in implicit expressions

Why is `diagnoseUnhandledAsyncSite` avoiding implicit expressions? I don’t know where exactly we currently introduce implicit expressions but could this be hiding other bugs as well where we’re not diagnosing an async call because it’s in an implicit expression. Could we remove the `isImplicit` check in `diagnoseUnhandledAsyncSite` instead?

Reply:
Yeah that’s definitely potentially hiding issues huh! If implicit code is violating isolation it’s still bad after all huh!

Reply:
So we're getting a bunch of double-diagnosed places with async let and similar friends in this situation... I'll see if I can fix that instead.

Reply:
FWIW unmarking this as implicit would also mean we could get rid of [this special case in MiscDiagnostics](https://github.com/apple/swift/blob/cd7e4c3db5f56ad51a7d51ec1a0d7ff245fb30a2/lib/Sema/MiscDiagnostics.cpp#L2301-L2331), so I'm not against unmarking it implicit if nothing else is depending on it being implicit. It's a bit of a weird case because although it is technically an implicit node (i.e generated by the compiler), it's also directly implied by what the user wrote, so could just as well be considered user-written code.

Reply:
Nice, removing weird special cases sounds good. I have to follow up in some more places it seems -- in ASTGen as well, in order to go down the `c` route but it seems promising

Reply:
Thanks for trying not checking for `isImplicit`, @ktoso. I agree that we can’t want to fix all the places that would generate duplicate diagnostics now, so not setting `isImplicit` (option c) seems like a reasonable solution to me.